packages/runtime/container-runtime/src/gc/garbageCollection.ts

```diff
@@ -350,115 +358,59 @@ export class GarbageCollector implements IGarbageCollector {
-			if (nodeData.unreferencedTimestampMs !== undefined) {
-				this.unreferencedNodesState.set(
-					nodeId,
-					new UnreferencedStateTracker(
-						nodeData.unreferencedTimestampMs,
-						this.configs.inactiveTimeoutMs,
-						currentReferenceTimestampMs,
-						this.configs.sweepTimeoutMs,
-						this.configs.sweepGracePeriodMs,
-					),
-				);
-			}
-			gcNodes[nodeId] = Array.from(nodeData.outboundRoutes);
+		// If the GC state has been initialized, update the tracking of unreferenced nodes as per the current
+		// reference timestamp.
+		for (const [, nodeStateTracker] of this.unreferencedNodesState) {
+			nodeStateTracker.updateTracking(currentReferenceTimestampMs);
 		}
-		this.gcDataFromLastRun = { gcNodes };
 	}
 
 	/**
 	 * Called when the connection state of the runtime changes, i.e., it connects or disconnects. GC subscribes to this
-	 * to initialize the base state for non-summarizer clients so that they can track inactive / sweep-ready nodes.
+	 * to initialize or update the unreference state tracking.
 	 * @param connected - Whether the runtime connected / disconnected.
 	 * @param clientId - The clientId of this runtime.
 	 */
 	public setConnectionState(connected: boolean, clientId?: string | undefined): void {
 		/**
-		 * For all clients, initialize the base state when the container becomes active, i.e., it transitions
-		 * to "write" mode. This will ensure that the container's own join op is processed and there is a recent
-		 * reference timestamp that will be used to update the state of unreferenced nodes. Also, all trailing ops which
-		 * could affect the GC state will have been processed.
-		 *
-		 * If GC is up-to-date for the client and the summarizing client, there will be an doubling of both
-		 * InactiveObject_Loaded and SweepReady_Loaded errors, as there will be one from the sending client and one from
-		 * the receiving summarizer client.
+		 * When the client connects (or reconnects), attempt to initialize or update the GC state. This will keep
+		 * the unreferenced state tracking updated as per the reference timestamp at the time of connection.
 		 *
-		 * Ideally, this initialization should only be done for summarizer client. However, we are currently rolling out
-		 * sweep in phases and we want to track when inactive and sweep-ready objects are used in any client.
+		 * During GC initialization and during connections in read mode, it is possible that either no ops are
+		 * processed or only trailing ops are processed. This means that the GC state is not initialized or initialized
+		 * with an older reference timestamp. So, doing this on every connection will keep the unreferenced state
+		 * tracking up-to-date.
 		 */
-		if (this.activeConnection() && this.configs.shouldRunGC) {
-			this.initializeGCStateFromBaseSnapshotP.catch((error) => {});
+		if (connected && this.configs.shouldRunGC) {
+			this.initializeOrUpdateGCState().catch((error) => {
```

Comment:
Could this happen _before_ all trailing ops have processed?  In the case where the `currentReferenceTimestampMs` is based on the `IConnect` message, I think that could cause some trouble.  Seems like that's not part of this PR, so it's safe, it's just using this to periodically reallign with the op stream that's been processed so far.  Does that sound right?

Reply:
Yes, this can happen before trailing ops are processed for read connections. I am not planning to use the `IConnected` timestamp so that should be fine. However, this means until we get a new op and it connects / reconnects, we will keep using old timestamps.

I don't think there ia way to get this 100% correct with what we currently have. Once we have the "reference" op, we could drop `Timers` and do the unreferenced state calculation based on the timestamp of the op for which GC is called, that should solve most if not all of the issues. Even in case of loads, we could use the timestamp of the last processed op to know what the state of the node is. Haven't thought through that one fully.

Reply:
Yeah need to give some more thought to how to drop the timers, that would be nice.  This approach makes sense for now, definite improvement and a simple model to reason over.